clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -3882,6 +3882,86 @@ def FMinOp : BinaryFPToFPBuiltinOp<"fmin", "MinNumOp">;
 def FModOp : BinaryFPToFPBuiltinOp<"fmod", "FRemOp">;
 def PowOp : BinaryFPToFPBuiltinOp<"pow", "PowOp">;
 
+//===----------------------------------------------------------------------===//
+// Assume Operations
+//===----------------------------------------------------------------------===//
+
+def AssumeOp : CIR_Op<"assume"> {
+  let summary = "Tell the optimizer that a boolean value is true";
```

Comment:
We need to mark this operation with some type of side-effect otherwise we'll make the situation with canonicalization worse (it will silently strip those away). We should probably have a simple CIR to CIR test using the canonicalizer and checking they don't get stripped away.

Looks like both `AssumeOp` and `AssumeSepStorageOp` need that (but perhaps not AssumeAlignedOp given the result is usually used?)

Reply:
Added a test in `clang/test/CIR/Transforms/builtin-assume.cir` but it seems that neither `-cir-simplify` nor `--canonicalize` erases these operations. Are there any other canonicalization passes that could do such optimization?

Reply:
That's interesting, I'd have expected these to be doing it (as they usually do for other operations), well at least there's a test to prevent that in case something off is going on in the meantime. Should be good for now, thanks for adding these.